zhusuan/distributions/univariate.py

```diff
@@ -1127,3 +1121,112 @@ def _log_prob(self, given):
+        See :class:`~zhusuan.distributions.base.Distribution` for more detailed
+        explanation.
+    :param is_reparameterized: A Bool. If True, gradients on samples from this
+        distribution are allowed to propagate into inputs, using the
+        reparametrization trick from (Kingma, 2013).
+    :param check_numerics: Bool. Whether to check numeric issues.
+    """
+
+    def __init__(self,
+                 temperature,
+                 logits,
+                 group_event_ndims=0,
+                 is_reparameterized=True,
+                 check_numerics=False):
+        self._logits = tf.convert_to_tensor(logits)
+        param_dtype = assert_same_float_dtype(
+            [(self._logits, 'BinConcrete.logits')])
+
+        self._temperature = assert_positive_integer(
+            temperature, param_dtype, 'BinConcrete.temperature')
+        if isinstance(self._temperature, (int, float)):
+            self._temperature = tf.constant(self._temperature, param_dtype)
+
+        self._check_numerics = check_numerics
+        super(BinConcrete, self).__init__(
+            dtype=param_dtype,
+            param_dtype=param_dtype,
+            is_continuous=True,
+            is_reparameterized=is_reparameterized,
+            group_event_ndims=group_event_ndims)
+
+    @property
+    def temperature(self):
+        """The temperature of BinConcrete."""
+        return self._temperature
+
+    @property
+    def logits(self):
+        """The log-odds of probabilities."""
+        return self._logits
+
+    def _value_shape(self):
+        return tf.constant([], dtype=tf.int32)
+
+    def _get_value_shape(self):
+        return tf.TensorShape([])
+
+    def _batch_shape(self):
+        return tf.shape(self.logits)
+
+    def _get_batch_shape(self):
+        return self.logits.get_shape()
+
+    def _sample(self, n_samples):
+        logits, temperature = self.logits, self.temperature
+        if not self.is_reparameterized:
+            logits = tf.stop_gradient(logits)
+            temperature = tf.stop_gradient(temperature)
+        shape = tf.concat([[n_samples], self.batch_shape], 0)
+
+        uniform = open_interval_standard_uniform(shape, self.dtype)
+        logistic = tf.log(uniform) - tf.log(1 - uniform)
+        samples = tf.sigmoid((logits + logistic) / temperature)
+
+        static_n_samples = n_samples if isinstance(n_samples, int) else None
+        samples.set_shape(
+            tf.TensorShape([static_n_samples]).concatenate(
+                self.get_batch_shape()))
+        return samples
+
+    def _log_prob(self, given):
+        temperature, logits = self.temperature, self.logits
+        log_given = tf.log(given)
+        log_1_minus_given = tf.log(1 - given)
+        log_temperature = tf.log(temperature)
+        logistic_given = log_given - log_1_minus_given
+
+        if self._check_numerics:
+            with tf.control_dependencies(
+                    [tf.check_numerics(log_given, "log(given)"),
```

Comment:
using `log_given = tf.check_numerics(log_given)` is simpler, but I don't understand why it is not right to use `control_dependencies` on `tf.identity`, and it seems that no reports about this can be found in Google?